testutil/fuse/fuse.go

```diff
@@ -0,0 +1,238 @@
+// +build linux darwin
+
+// Copyright 2019 The qiffang Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"fmt"
+	"github.com/prometheus/tsdb/testutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+)
+
+type HookFs struct {
+	Original   string
+	Mountpoint string
+	FsName     string
+	fs         pathfs.FileSystem
+	hook       Hook
+}
+
+func NewHookFs(original string, mountpoint string, hook Hook) (*HookFs, error) {
+	loopbackfs := pathfs.NewLoopbackFileSystem(original)
+	hookfs := &HookFs{
+		Original:   original,
+		Mountpoint: mountpoint,
+		FsName:     "hookfs",
+		fs:         loopbackfs,
+		hook:       hook,
+	}
+	return hookfs, nil
+}
+
+func (h *HookFs) String() string {
+	return fmt.Sprintf("HookFs{Original=%s, Mountpoint=%s, FsName=%s, Underlying fs=%s, hook=%s}",
+		h.Original, h.Mountpoint, h.FsName, h.fs.String(), h.hook)
+}
+
+func (h *HookFs) SetDebug(debug bool) {
```

Comment:
I just tried it and you can avoid all this extra boilerplate by using struct embedding like:

```
type LoopBackFs struct {
	pathfs.FileSystem
}

type HookFs struct {
	Original   string
	Mountpoint string
	FsName     string
	LoopBackFs
	hook Hook
}

func NewHookFs(original string, mountpoint string, hook Hook) (*HookFs, error) {
	hookfs := &HookFs{
		Original:   original,
		Mountpoint: mountpoint,
		FsName:     "hookfs",
		LoopBackFs: LoopBackFs{pathfs.NewLoopbackFileSystem(original)},
		hook:       hook,
	}
	return hookfs, nil
}
```

Reply:
Thanks for your comments, and updated it.
But for now, there is only rename which has hook, actually almost all methods should have hook to inject all filesystem error.
So if almost methods have the hook like rename method, there will be also `LoopBackFs` boilerplate.